### PR TITLE
Refactor iterator parsing to allow more reuse

### PR DESCRIPTION
**Pull request by ephphatha**
Trying to avoid repetitive code when loading data from more external files (playerData is likely going to be split into three files for attributes, loadout, equipment bonuses for example...)

I'm not particularly happy with how the iterators are defined, the initial aim was to remove the wrapper so range-based loops are more direct but this makes it harder to keep track of the column number for error reporting. Might need to reintroduce wrappers. I also think the definition of `RecordIterator::operator*` is awful and it's likely a terrible idea 😂 . Hoping someone better at iterators than me (glebm? 😉 ) can suggest improvements/a better refactor.

I also need to add more tests around repeated calls to parseInt/value access on a FieldIterator.

### CODE REVIEW

File: Source/data/iterators.hpp [hunk 1]
```diff
@@ -5,230 +5,251 @@
...
 		}
+		return *this;
+	}
 
-		[[nodiscard]] value_type operator*()
-		{
-			return { &state_, end_ };
-		}
-	};
+	[[nodiscard]] reference_type operator*()
```
Comment (glebm) [inline Source/data/iterators.hpp:215]: This suggests that `RecordIterator` should not be an iterator.

Perhaps the following structure:

* `RecordsRange`: `begin()/end() RecordIterator`
* `RecordIterator`: `value_type = FieldsRange`, `row()` method
* `FieldsRange`: `begin() / end() FieldIterator`, also `row()` method here
* `FieldIterator`: `value_type = std::string_view`, `column()` method

Then the iteration would be as follows:

```c++
for (FieldsRange fields : dataFile.records()) {
  FieldIterator it = fields.begin();
  FieldIterator end = fields.end();
  
  // fields.row(),  it.column()
}
```

File: Source/data/iterators.hpp [hunk 2]
```diff
@@ -1,234 +1,415 @@
...
+	* This requires an actual non-const value access to happen first before it returns
+	* any useful results, intended for use in error reporting (or test output).
+	*/
+	[[nodiscard]] std::string_view currentValue() const
+	{
+		return state_->value;
+	}
+};
+
+#ifdef BUILD_TESTING
```
Comment (glebm) [inline Source/data/iterators.hpp:149]: This can be defined in the test itself or just always defined (LTO will remove unused code)

Comment (ephphatha) [reply Source/data/iterators.hpp]: Thanks, I'm definitely glad to remove the `#ifdef` guard :D (guess we should do the same for `PointOf` and friends as well)

File: Source/data/iterators.hpp (hunk 1 above)
Comment (ephphatha) [reply Source/data/iterators.hpp]: I ended up going with a slightly different approach, the RecordsRange/RangeIterator/FieldsRange/FieldIterator structure was the thing I wasn't really happy with.

Now a DataFile is a range of Records (using a RecordIterator as a forward_iterator). A Record is a range of Fields (using a FieldIterator as a forward_iterator).

This lets a simple loop over a file be written as:
```cpp
DataFile file;
for (Record record : file) {
  for (Field field : record) { // alternatively `for (std::string_view field : record) {}` can be used if the value will only ever be used as a string and row/column numbers don't matter
    if (field == "string_view"sv)
      // can treat the value as if it was a string_view
    // can use field.parseInt(dest) or expected = field.asInt<int8_t>() to get a numeric value
    // field.row() and field.column() return the current file position
  }
}
```

Having the row/column available at the field level turned out to be really helpful for error messages
2023-08-28T233236_devenv

File: Source/data/iterators.hpp [hunk 3]
```diff
@@ -1,234 +1,408 @@
...
-		{
-			if (state_ == nullptr && rhs.state_ == nullptr)
-				return true;
+	/**
+	 * Allows using a field transparently as a string_view
+	 */
+	[[nodiscard]] operator std::string_view()
+	{
+		return this->value();
+	}
```
Comment (glebm) [inline Source/data/iterators.hpp:57]: Implicit conversions are a feature of C++ best avoided most of the time. Let's remove this conversion and be explicit when we need a `string_view` (it's just a `*` after all). https://google.github.io/styleguide/cppguide.html#Implicit_Conversions

File: Source/data/iterators.hpp [hunk 4]
```diff
@@ -1,234 +1,408 @@
...
+	 */
+	[[nodiscard]] unsigned column() const
+	{
+		return column_;
 	}
 };
 
-class RecordsRange {
-	const char *const begin_;
+class Record {
```
Comment (glebm) [inline Source/data/iterators.hpp:286]: The names `Record` and `Field` in the `devilution` namespace are too general.
How about `DataFileRecord`, `DataFileField`?

File: test/data_file_test.cpp [hunk 5]
```diff
@@ -144,12 +145,49 @@ TEST(DataFileTest, IterateOverRecords)
...
+	auto result = LoadDataFile("txtdata\\lf.tsv");
+	ASSERT_TRUE(result.has_value()) << "Unable to load lf.tsv";
+
+	DataFile &dataFile = result.value();
+
+	std::vector<std::vector<std::string_view>> expectedFields {
+		{ "", "2", "3" },
+		{ "1", "2", "" },
+		{ "1", "", "3" },
+		{ "" } // file ends with a newline, parsing returns a single empty field
```
Comment (glebm) [inline test/data_file_test.cpp:175]: Text files generally end with a newline, how come it is parsed as an empty field?

Shouldn't the following:

```
"a\nb\n"
```

Parse as: `{ "a" }, { "b" }`?

File: Source/data/iterators.hpp (hunk 1 above)
Comment (glebm) [reply Source/data/iterators.hpp]: Looks great, except for the implicit `string_view` conversion (left a comment) 🎉

File: Source/data/iterators.hpp (hunk 4 above)
Comment (ephphatha) [reply Source/data/iterators.hpp:286]: Definitely agree there. Would it be reasonable to declare them as inner classes of `DataFile` so they get accessed as `DataFile::Record` etc? Or introduce another namespace like `devilution::textdata::Record`?

Comment (glebm) [reply Source/data/iterators.hpp:286]: Yeah either of these options are fine

File: test/data_file_test.cpp (hunk 5 above)
Comment (ephphatha) [reply test/data_file_test.cpp:175]: That was initially to remove ambiguity, given inconsistencies between the way various programs produce tab-delimited files e.g. google sheets doesn't include a trailing newline on tsv exports). It seemed easiest to use newlines as separators at the time of #6399.

It's relatively trivial to change and I can make the behaviour less surprising by documenting the format as a collection of records terminated (not separated) by `\n` (with the last record having an optional terminator). Means I can get rid of the uses of the `Field::endOfFile` method too.

File: Source/data/iterators.hpp (hunk 3 above)
Comment (ephphatha) [reply Source/data/iterators.hpp]: removed the operator==(std::string_view) definition as well since it had a similar effect.

**Review by glebm — DISMISSED** [review]

File: test/data_file_test.cpp [hunk 6]
```diff
@@ -103,12 +102,78 @@ TEST(DataFileTest, LoadTypicalFile)
...
+	case DataFile::Error::BadRead:
+		return "cannot read contents";
+	default:
+		return "unexpected error";
+	}
+}
+
+TEST(DataFileTest, LoadEmptyFile)
+{
+	//GTEST_SKIP() << "Loading 0 length assets is not supported by AssetHandle, trying to load this file fails with Error::BadRead";
```
Comment (AJenbo) [inline test/data_file_test.cpp:144]: I'm just guessing here:
```suggestion
	// GTEST_SKIP() << "Loading 0 length assets is not supported by AssetHandle, trying to load this file fails with Error::BadRead";
```

Comment (ephphatha) [reply test/data_file_test.cpp:144]: Yeah this needs to be removed, I added a workaround that meant I no longer need to skip this test but forgot to delete the comment/line. Got sidetracked trying to fix the grammar.

Comment (AJenbo) [reply test/data_file_test.cpp:144]:
```suggestion
```

**Pull request merged**